src/performance/strategies.py

```
@@ -0,0 +1,179 @@
+    def optimize_memory_usage(self, func: Callable) -> Callable:
+        """Optimize memory-intensive operations"""
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            if self._is_memory_constrained():
+                return self._apply_memory_optimization(func, *args, **kwargs)
+            return func(*args, **kwargs)
+        return wrapper
+        
+    def _is_cpu_constrained(self) -> bool:
+        """Check if CPU usage is above threshold"""
+        return psutil.cpu_percent() / 100 > self.cpu_threshold
```

Ruff (0.8.2)

Undefined name `psutil`

(F821)